go/package/download.php

```diff
@@ -0,0 +1,152 @@
+        exit;
+      }
+
+      if(empty($version)) {
+        // If version isn't provided, we'll query the live api for the version
+        $json = @file_get_contents(KeymanHosts::Instance()->api_keyman_com . '/keyboard/' . rawurlencode($id));
+        if ($json !== FALSE) {
+          $json = json_decode($json);
+        }
+
+        if(empty($json)) {
+          echo "Invalid parameter; id not found";
+          exit;
+        }
+
+        $version = $json->version;
+      }
+
+      $update = empty($update) ? 0 : 1;
+
+      $bcp47 = Validation::validate_bcp47($bcp47);
+      $tier = Validation::validate_tier($tier, 'stable');
+      $platform = Validation::validate_platform($platform, 'unknown');
+
+      $url = KeymanHosts::Instance()->downloads_keyman_com . "/keyboards/$id/$version/$id.kmp";
+      $url_e = htmlentities($url);
+
+      if(DEBUG) header('Content-Type: text/plain');
+
+      self::report_download_event($ga_cookie, $id, $platform, $tier, $bcp47, $update);
+
+      if(DEBUG) {
+        echo "\n\nLocation: $url\n";
+        exit;
+      }
+
+      if(KeymanHosts::Instance()->Tier() !== KeymanHosts::TIER_TEST) {
+        // We don't do a redirect for Test tier because a test instance of the
+        // downloads server is not available and so it gives us an error
+        header("HTTP/1.1 302 Found");
+        header("Cache-Control: no-store");
+        header("Location: $url");
+      }
+
+      echo "<a href='$url_e'>Download Link</a>";
+    }
+
+    private static function report_download_event($ga_cookie, $id, $platform, $tier, $bcp47, $update) {
+      $cid = '';
+      if(empty($ga_cookie)) $ga_cookie = '';
+
+      if(preg_match('/^GA\d+\.\d\.(.+)$/', $ga_cookie, $matches)) {
+        $cid = rawurlencode($matches[1]);
+        $gauser = "cid={$cid}";
+      } else {
+        $uid = GUIDv4(true);
+        $gauser = "uid={$uid}";
+      }
+
+      if(DEBUG) {
+        $gabaseurl = "https://www.google-analytics.com/debug/collect";
+      } else {
+        $gabaseurl = "https://www.google-analytics.com/collect";
+      }
+
+      $update = $update ? "update" : "install";
+
+      $category = 'keyboard';
+      $action = rawurlencode("download-$platform-$tier-$update" . (empty($bcp47) ? "" : "-$bcp47"));
+      $label = rawurlencode($id);
+
+      $gaurl = "$gabaseurl?v=1&t=event&tid=UA-249828-1&$gauser&ds=server&ec=$category&ea=$action&el=$label";
+      $result = @file_get_contents($gaurl);
+      if(DEBUG) {
+        var_dump("Google Analytics response for $gaurl: ".print_r($http_response_header, true));
+      }
+
+      return !($result === FALSE);
+    }
+  }
+
+/**
+* Returns a GUIDv4 string
+*
+* Uses the best cryptographically secure method
+* for all supported pltforms with fallback to an older,
+* less secure version.
+* https://www.php.net/manual/en/function.com-create-guid.php#119168
+*
+* @param bool $trim
+* @return string
+*/
+function GUIDv4 ($trim = true)
```

Comment:
I checked how [web-languageforge](https://github.com/sillsdev/web-languageforge) generates guids. They get https://github.com/ramsey/uuid from 
```
composer require ramsey/uuid
```

Reply:
If you stick with GUIDv4, maybe split to a separate file?

Reply:
For now, this is a fallback because some servers may not have `com_create_guid()`. I can move to a separate file but I will leave that for a future refactor I think (I am guessing that we'll always be using `com_create_guid()` on the live machines).